Add a wrapper to transform a NumPy function into a PyTorch function

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #114610

A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

lezcano committed:
Add a wrapper to transform a NumPy function into a PyTorch function

A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

[ghstack-poisoned]

lezcano committed:
Update on "Add a wrapper to transform a NumPy function into a PyTorch function"


A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano committed:
Update on "Add a wrapper to transform a NumPy function into a PyTorch function"


A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano committed:
Update on "Add a wrapper to transform a NumPy function into a PyTorch function"


A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on docs/source/torch.compiler_faq.rst:
```diff
@@ -614,23 +614,58 @@ to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
+.. code-block:: python
+
+   @torch.compile
+   @torch.compiler.wrap_fn
```
Should this (and below) be `wrap_np`?  Doesn't seem to match code.

jansel commented on torch/_dynamo/external_utils.py:
```diff
@@ -27,3 +35,21 @@ def call_hook(hook, *args):
     return result
+
+
+def wrap_np(f):
```
Should we call this wrap_numpy?  Or something more descriptive?  Wrap is pretty generic...

lezcano replied:
Renamed to the more explicit `wrap_numpy`

lezcano committed:
Update on "Add a wrapper to transform a NumPy function into a PyTorch function"


A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel reviewed (commented):
Code changes look good to me, though I'd like others to weigh in on if this should be a public API and what it should be called.

albanD commented:
I'm not sure the semantic of this transformation is very clear to me and when I should use it?

Maybe adding type annotation to the functions will make it easier to see what happens?

lezcano commented:
The type annotation would be `Callable[Any, Any] -> Callable[Any, Any]`, so I don't think it would add all that much...

albanD commented:
I mean in terms of Tensor vs array.

lezcano committed:
Update on "Add a wrapper to transform a NumPy function into a PyTorch function"


A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented:
Sorry, I completely dropped the ball on this one. Added type annotations to all functions. It's not 100% clear to me whether this makes things simpler or not, given that the decorator changes the types, but yeah...

albanD commented on docs/source/torch.compiler_faq.rst:
```diff
@@ -614,23 +614,58 @@ to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
+on CPU. Note also that calling ``.numpy(force=True`` under ``torch.compile`` also
+moves to CPU and deactivates gradient computations.
+
+We provide the decorator ``torch.compiler.wrap_numpy``  that implements the patteron of wrapping
```
I would move this at the top actually. Users should use this instead of doing it by hand in most cases?

albanD commented on torch/compiler/__init__.py:
```diff
@@ -118,3 +119,31 @@ def rand_foo():
+    It is designed to be used with :func:`torch.compile`. It allows to compile a NumPy function
+    as if it were a PyTorch function. If the function can be traced without graph breaks,
+    this allows you to run NumPy code on CUDA.
+    Similarly, you can pass Tensors with ``requires_grad=True`` and then call ``backward()`` on the output
```
nit to shorten the doc, this can be folded in the sentence above: `code on CUDA and with gradients enabled.`

albanD commented on docs/source/torch.compiler_faq.rst:
```diff
@@ -614,23 +614,58 @@ to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
+
+   X = torch.randn(1024, 64, device="cuda", requires_grad=True)
+   Y = torch.randn(1024, 64, device="cuda")
+   with torch.device("cuda"):
```
The context manager is not needed here actually? If so I think we should remove it from here.

lezcano replied:
The context manager allows for graphbreaks to happen in the compiled code. Let me discuss graph breaks in the FAQ as well.

albanD reviewed (commented):
I think that there is a pretty big constraint here that we end up with a function that can ONLY be used in the context of compile-full-graph and will NOT work in eager.
This is a new enough concept that I think we should say it explicitly in the faq and docstring.

lezcano committed:
Update on "Add a wrapper to transform a NumPy function into a PyTorch function"


A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented:
Addressed the review. In particular:
- I have streamlined the examples in the FAQ, using only `wrap_numpy` as it's the preferred approach 
- I have added a paragraph discussing graph breaks explicitly.
- I have added a note discussing graph breaks explicitly to the docs.
  - Note that the docs are not currently rendered, as the `torch.compiler` page of the docs is currently being used as a page linking to other pages.

albanD commented on docs/source/torch.compiler_faq.rst:
```diff
@@ -602,35 +602,84 @@ context. Consider the example
 to avoid all these rather expensive memory copies. To do so, we just need
-to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
+to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors.
+We can do so by using ``torch.compiler.wrap_numpy``:
```
Thanks for the update, it looks quite good.
Should this piece be split into two questions in the FAQ?
- How to speed up NumPy code with `torch.compile`?
  - Use the fast cpu backend by enabling torch.compile
  - Run things on CUDA
- How to use NumPy code within your PyTorch code?
  - The wrap_numpy function

lezcano replied:
I think splitting it is too much given how little information there is. I will add one example where we simply run things on CPU, implementing your first point "Use the fast cpu backend by enabling torch.compile"

lezcano replied:
Nevermind, after writing it I chose to split it into two sections, although they are different to the ones you proposed.

lezcano committed:
Update on "Add a wrapper to transform a NumPy function into a PyTorch function"


A less general version of this wrapper was used in the keynote on
`torch.compile(numpy)`. We expose a generic version of the wrapper
that works seamlessly with `torch.compile`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

albanD reviewed (approved):
Thanks for the update!

lezcano commented:
@pytorchbot merge
